api/layer/object.go

```diff
@@ -252,31 +259,61 @@ func (n *layer) ListObjectsV2(ctx context.Context, p *ListObjectsParamsV2) (*Lis
 		err        error
 		result     ListObjectsInfoV2
 		allObjects []*ObjectInfo
+		bkt        *BucketInfo
+		accessKey  string
+		cacheKey   string
 	)
 
 	if p.MaxKeys == 0 {
 		return &result, nil
 	}
 
-	if p.ContinuationToken != "" {
-		// find cache with continuation token
+	if bkt, err = n.GetBucketInfo(ctx, p.Bucket); err != nil {
+		return nil, err
+	}
+
+	if data, ok := ctx.Value(api.BoxData).(*accessbox.Box); ok && data != nil {
+		accessKey = data.Gate.AccessKey
 	} else {
+		return nil, fmt.Errorf("couldn't get access key from the accessbox: %s", err)
+	}
+
+	cacheKey = createKey(accessKey, bkt.CID)
+
+	if p.ContinuationToken != "" {
+		allObjects = n.cache.GetCache(p.ContinuationToken, cacheKey)
+		allObjects = trimStartAfter(p.StartAfter, allObjects)
+	}
+
+	if allObjects == nil {
 		allObjects, err = n.listSortedAllObjects(ctx, allObjectParams{
-			Bucket:     p.Bucket,
+			Bucket:     bkt,
 			Prefix:     p.Prefix,
 			Delimiter:  p.Delimiter,
 			StartAfter: p.StartAfter,
 		})
 		if err != nil {
 			return nil, err
 		}
+
+		if p.ContinuationToken != "" {
+			for i, obj := range allObjects {
```

Comment:
I supposes we should extract this:
```
for i, obj := range allObjects {
	if obj.ID().String() == p.ContinuationToken {
		allObjects = allObjects[i:]
		break
	}
}
``` 
to separate function and use it here and in `GetCache`

Reply:
Didn't notice repeated code, fixed